### DISCUSSION TITLE
Visualize hierarchical logging/distributed tracing

### DISCUSSION THREAD

**Discussion by ngbrown** [datalust/seq-tickets#1105] [Ideas]
My services include [hierarchical RequestId](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/HierarchicalRequestId.md) information.  I would like to visualize it with a timeline or flame chart.
An example is https://www.jaegertracing.io/docs/1.10/

I don't know if plugins are allowed to make changes to the UI. If they are, then this would be a good plugin.

**Comment by nblumhardt** [comment]
Thanks :+1: - I think it's time we investigated/started assembling a position or plan-of-attack for hierarchical request ids/tracing.

**Comment by nblumhardt** [comment]
Note #1011, which calls attention to OpenTelemetry.

**Comment by ngbrown** [comment]
Since I created this issue, .NET Core 3.0/3.1 has now added support for more formats, namely `ActivityIdFormat.W3C`. Evidently this will be [the default format for .NET 5.0](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.defaultidformat).  I still use `ActivityIdFormat.Hierarchical` in my services.

When the request comes from the browser, I generate/rotate a random root activity id every time the displayed url changes.  This is then passed on in any requests to the back end and all the services add to the hierarchy for any nested requests.

When I want to trace a stream in my logs, starting from a log entry of interest, I click on the green filter checkmark for the `RequestId`, then I have to manually edit the query to use a `LIKE` and the wildcard (%), and run the query again:

```
RequestId LIKE '|e94fd8c4-42891a2d35276580.%'
```

Ideally this feature would do all that automatically.  It would automatically find the root activity id, start the top level query, and then show a timeline graph.

It would show the whole hierarchy of requests in a timeline sort of like the overview pane of the network tab in Chrome Dev tools.  Or a flame graph (like on Chrome's performance tab) if that's better.  The horizontal axis would be time (with at least relative ms/seconds displayed). The vertical axis would be top level parent activity Id at the top, and each layer down would be a nested activity id from any service feeding it's logs into Seq.

Then, I can click on any segment of the graph, and see everything from that level down. (Again, using the LIKE and wildcard in the correct level).  Maybe a Ctrl+Click on a segment would filter to only that service, not including the nested calls.

For colors on the bars, I think re-using the same color for the same service (I use `ApplicationName`) would be helpful.  If there are any errors or warnings, the graph could show warning/error icons or colors on those sections automatically.

I haven't used any of the other products available on the market for hierarchical RequestId or OpenTelemetry, but the above is what I would expect out of that sort of feature.

**Comment by milang** [comment]
@nblumhardt any updates on hierarchical call visualization, OpenTelemetry support in Seq?

Would it be at least possible to publish a guidance document for using Serilog/Seq in distributed applications (e.g. recommendations for writing Serilog logging code that is [`System.Diagnostics.Activity`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity)-"friendly" + prepares today's code for future OpenTelemetry Seq functionality, if any) ?

Have you added any distributed tracing functionality to Seq implementation while developing high availability / clustering support? Any lessons learned you would be willing to share with your Seq community?

**Comment by nblumhardt** [comment]
Thanks for the ping, @milang! Agreed that a how-to on this is sorely needed. We're putting the final touches on 2021.3; I've added this to our list of posts/documentation to consider after the new version has shipped.

Seq's internal (storage engine) logging isn't a great example for this, since the processes/topology are tightly fixed. We lean very heavily on metrics for low-level instrumentation, and only tend to use rich logging at the "application layer". Probably also a good blog post for us to consider :-)

**Reply by a98c14** [reply]
Hi, any updates?

**Comment by nblumhardt** [comment]
Hi all! Seq 2023.2 will add OTLP log ingestion, we're tracking it in https://github.com/datalust/seq-tickets/issues/1679. This is the first step in a multi-part process to bring more of this functionality to Seq. Feedback greatly appreciated, if you're able to check out our progress/current preview release.

**Comment by nblumhardt** [comment]
Seq now fully supports hierarchical, distributed traces 🎉

Both OpenTelemetry/OTLP, and direct HTTP/JSON ingestion (e.g. using SerilogTracing) are supported. See https://docs.datalust.co/docs/tracing for all of the details.

Thanks for all of your feedback and encouragement, we're really happy with the way this all came together! 😁

**Reply by cilerler** [reply]
Does the docker image [2024.1.10981](https://hub.docker.com/r/datalust/seq) support this feature?

**Reply by liammclennan** [reply]
Hi @cilerler 

Yes it does.
